### ISSUE TITLE
ServiceAccount incompatability with EKS 1.21+

### ISSUE THREAD

**Issue by tim-sendible** [hashicorp/vault#15390]
**Describe the bug**
Apologies if this is the wrong place. I received the following alert from AWS, and I think this indicates that Vault is not using a newer version of the Kubernetes client SDK, which will result in ServiceAccounts not working fairly shortly:

```
We have identified applications running in one or more of your Amazon EKS clusters that are not refreshing service account tokens. Applications making requests to Kubernetes API server with expired tokens will fail. You can resolve the issue by updating your application and its dependencies to use newer versions of Kubernetes client SDK that automatically refreshes the tokens.

What is the problem?

Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].

To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.

How can you resolve the issue?

To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:

* Go v0.15.7 and later
* Python v12.0.0 and later
* Java v9.0.0 and later
* Javascript v0.10.3 and later
* Ruby master branch
* Haskell v0.3.0.0

We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version. 
```

**To Reproduce**
1. Deploy Vault into EKS 1.22 (we used the official vault helm chart)
2. Ensure that serviceaccount.create is 'true' and that an eks role-arn annotation is added to it.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.10.2
* Vault CLI Version (retrieve with `vault version`): n/a
* Server Operating System/Architecture: EKS 1.22

**Comment by rodosskiand** [comment]
https://www.vaultproject.io/docs/auth/kubernetes

`disable_iss_validation: true` is a workaround?

**Comment by tim-sendible** [comment]
> `disable_iss_validation: true` is a workaround?

This isn't related to the kubernetes auth method. This is related to the serviceAccount that Vault uses to interact with other aws components.

**Comment by rodosskiand** [comment]
> > `disable_iss_validation: true` is a workaround?
> 
> This isn't related to the Kubernetes auth method. This is related to the serviceAccount that Vault uses to interact with other AWS components.

Wich resources do you mean? I thought that service accounts are used for communication with kubernetes api.

**Comment by hamza-alzaatreh** [comment]
@rodosskiand we have received the same email from AWS I believe its the vault Kubernetes service account that used to interact with the Kubernetes API not the vault Kubernetes auth method.

**Comment by nicolaelapteanu** [comment]
Hello,

I have also received a similar email from AWS.

A piece of the message looks like is (cluster name, account and region was masked):
```
As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens. Service accounts are listed in the format : <eks-cluster-arn>|<namespace:serviceaccount>

arn:aws:eks:<region>:<account>:cluster/<cluserName>|vault:vault
```
As per AWS [documentation](https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#identify-pods-using-stale-tokens), I have enabled Control plane logging on a couple of the EKS clusters where Vault is running and have run the Cloudwatch query. I could not see anything there, but that might be related to the fact that we're running Vault cluster on AWS Spot Instances and it can get a few restarts per day.

After a simple checking of the Vault source code, I see that the Kubernetes SDK version used on Vault is [0.12](https://github.com/hashicorp/vault/blob/main/go.mod#L102) though Amazon recommends [v0.15.7](https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#identify-pods-using-stale-tokens) and later.

Hope this helps to determine if that's a Kubernetes SDK deprecation notice, or smth else.

Thanks!

**Comment by tsaarni** [comment]
@rodosskiand is right, the error described by @tim-sendible is about service account token that Vault uses to communicate with Kubernetes  API. 

You can read about the short-lived tokens [here](https://www.vaultproject.io/docs/auth/kubernetes#kubernetes-1-21) and to fix the problem follow the instructions [here](https://www.vaultproject.io/docs/auth/kubernetes#use-local-service-account-token-as-the-reviewer-jwt).  Helm [configuration example](https://www.vaultproject.io/docs/platform/k8s/helm/examples/kubernetes-auth) has not been updated for this.

**Comment by tstraley** [comment]
@tsaarni  I'm confused by this, since the docs say Vault will periodically re-read the file to support short-lived tokens, [yet the code here](https://github.com/hashicorp/vault/blob/main/api/auth/kubernetes/kubernetes.go) seems to only read the file once when initializing NewKubernetesAuth.

**Comment by tsaarni** [comment]
> @tsaarni I'm confused by this, since the docs say Vault will periodically re-read the file to support short-lived tokens, [yet the code here](https://github.com/hashicorp/vault/blob/main/api/auth/kubernetes/kubernetes.go) seems to only read the file once when initializing NewKubernetesAuth.

I was referring to the Kubernetes auth method code here https://github.com/hashicorp/vault-plugin-auth-kubernetes. Vault itself uses this code to authenticate towards Kubernetes.

The code you linked is the API that Go based applications can use to authenticate towards Vault ([mentioned here](https://www.vaultproject.io/docs/auth/kubernetes#code-example)). You are right, it  reads the token once, so maybe users of this API are impacted as well. I have not used the Go API myself. I'm unsure if clients would reuse the config when they need to reauthenticate (I think only after connection has failed). Interesting question!

**Comment by tomhjp** [comment]
@tsaarni is correct with the above (thank you!), and I'm not aware of any more documentation locations with outdated instructions once #15675 is merged, so I'll close this for now.

To summarise, this issue is most commonly resolved by using Vault 1.9.3 or later when Vault is inside Kubernetes, and _not_ setting the `token_reviewer_jwt` parameter when writing to the k8s auth config endpoint. Vault will then automatically read in and use new tokens as they're refreshed. There are other options enumerated [here](https://www.vaultproject.io/docs/auth/kubernetes#kubernetes-1-21), including JWT auth and long-lived tokens, for Vault living outside the k8s cluster.
